src/bls12_381/fp2.rs

```diff
@@ -0,0 +1,1236 @@
+        for i in (2..=C1).rev() {
+            let tv5 = i as u32 - 2; // 18.    tv5 = i - 2
+            let tv5 = num_bigint::BigUint::from(2u64).pow(tv5); // 19.    tv5 = 2^tv5
+            let tv5 = tv4.pow_vartime(&tv5.to_u64_digits()); // 20.    tv5 = tv4^tv5
+            let e1 = tv5.ct_eq(&Fp2::one()); // 21.    e1 = tv5 == 1
+            let tv2 = tv3 * tv1; // 22.    tv2 = tv3 * tv1
+            tv1 = tv1 * tv1; // 23.    tv1 = tv1 * tv1
+            let tv5 = tv4 * tv1; // 24.    tv5 = tv4 * tv1
+            tv3.conditional_assign(&tv2, !e1); // 25.    tv3 = CMOV(tv2, tv3, e1)
+            tv4.conditional_assign(&tv5, !e1); // 26.    tv4 = CMOV(tv5, tv4, e1)
+        }
+        (is_square, tv3)
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct Fp2Bytes {
+    pub slice: [u8; 96],
+}
+
+impl Default for Fp2Bytes {
+    fn default() -> Self {
+        Self { slice: [0u8; 96] }
+    }
+}
+
+impl AsMut<[u8]> for Fp2Bytes {
+    fn as_mut(&mut self) -> &mut [u8] {
+        &mut self.slice
+    }
+}
+
+impl AsRef<[u8]> for Fp2Bytes {
+    fn as_ref(&self) -> &[u8] {
+        &self.slice
+    }
+}
+
+impl ff::PrimeField for Fp2 {
+    type Repr = Fp2Bytes;
+
+    const NUM_BITS: u32 = 381;
+    const CAPACITY: u32 = 381 - 1;
+    const MODULUS: &'static str =
+        "0x1a0111ea397fe69a4b1ba7b6434bacd764774b84f38512bf6730d2a0f6b0f6241eabfffeb153ffffb9feffffffffaaab";
+    const MULTIPLICATIVE_GENERATOR: Self = unimplemented!();
+    const ROOT_OF_UNITY: Self = Self {
+        c0: Fp::from_raw_unchecked([
+            0x7bcf_a7a2_5aa3_0fda,
+            0xdc17_dec1_2a92_7e7c,
+            0x2f08_8dd8_6b4e_bef1,
+            0xd1ca_2087_da74_d4a7,
+            0x2da2_5966_96ce_bc1d,
+            0x0e2b_7eed_bbfd_87d2,
+        ]),
+        c1: Fp::from_raw_unchecked([
+            0x7bcf_a7a2_5aa3_0fda,
+            0xdc17_dec1_2a92_7e7c,
+            0x2f08_8dd8_6b4e_bef1,
+            0xd1ca_2087_da74_d4a7,
+            0x2da2_5966_96ce_bc1d,
+            0x0e2b_7eed_bbfd_87d2,
+        ]),
+    };
+    const ROOT_OF_UNITY_INV: Self = Fp2 {
+        c0: Fp::zero(),
+        c1: Fp::zero(),
+    };
+    const DELTA: Self = Fp2 {
+        c0: Fp::zero(),
+        c1: Fp::zero(),
+    };
+    const TWO_INV: Fp2 = Fp2 {
+        c0: Fp::zero(),
+        c1: Fp::zero(),
+    };
+    const S: u32 = 0;
+
+    fn from_repr(r: Self::Repr) -> CtOption<Self> {
+        // This uses little endian and so, assumes the array passed in is
+        // in that format.
+        Self::from_bytes(&r.slice)
+    }
+
+    fn to_repr(&self) -> Self::Repr {
+        Fp2Bytes { slice: self.to_bytes() }
+    }
+
+    fn is_odd(&self) -> Choice {
+        Choice::from(self.to_bytes()[0] & 1)
+    }
+}
+
+impl From<u64> for Fp2 {
+    fn from(val: u64) -> Self {
+        Fp2 {
+            c0: Fp::from(val),
+            c1: Fp::zero(),
+        }
+    }
+}
+
+impl WithSmallOrderMulGroup<3> for Fp2 {
```

Comment:
Is this trait needed? because right now `ZETA` is not a 3rd root of unity

Reply:
`CurveExt` trait requires it https://github.com/zcash/pasta_curves/blob/main/src/arithmetic/curves.rs#L32 since `Fp2` is used as base field for `G2Affine`